[Merged by Bors] - feat(algebra/big_operators/finprod): add lemma `finprod_mem_finset_of_product`

ocfnash opened the pull request:
---

Related conversation on Zulip [here](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/finsum.20over.20product.20of.20types/near/236961309)

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/big_operators/finprod): add lemma `finprod_mem_finset_of_product`

ocfnash committed:
Appease classical linter

eric-wieser commented on src/algebra/big_operators/finprod.lean:
```diff
@@ -700,4 +700,33 @@ lemma finsum_mul {R : Type*} [semiring R] (f : α → R) (r : R)
+    (s.mul_support_of_fiberwise_prod_subset_image f prod.fst),
+    ← finset.prod_fiberwise_of_maps_to _ f],
+  finish,
+end
```
I don't think this is a particularly good lemma, as I can't chain it to handle a product of products. A better lemma IMO might be something like:
```lean
@[to_additive] lemma finprod_curry  (f : α × β → M) (h : (mul_support f).finite) :
  ∏ᶠ ab, f ab = ∏ᶠ a b, f (a, b) :=sorry
```
from which you can recover this lemma as hopefully a tidied version of:
```lean
@[to_additive] lemma finprod_mem_finset_of_product (s : finset (α × β)) (f : α × β → M) :
  ∏ᶠ ab (h : ab ∈ s), f ab = ∏ᶠ a b (h : (a, b) ∈ s), f (a, b) :=
begin
  rw finprod_curry,
  classical,
  have : (λ (ab : α × β), ∏ᶠ (h : ab ∈ s), f ab) = (λ ab, if ab ∈ s then f ab else 1),
  { ext,
    split_ifs,
    rw [eq_true_intro h, finprod_true],  -- there should be a better lemma for this!
    rw [eq_false_intro h, finprod_false], },
  rw this,
  rw mul_support,
  apply (set.finite_mem_finset s).subset,
  intros x hx,
  simp at ⊢ hx,
  exact hx.1,
end
```

ocfnash replied:
Hmm, good point and I agree; `finprod_curry` is certainly worth having. I'll put together a proof this afternoon.

I wonder if the following might also be desirable:
```lean
@[to_additive] lemma finprod_mem_finset_of_product' [decidable_eq α] [decidable_eq β]
  (s : finset (α × β)) (f : α × β → M) :
  ∏ᶠ ab (h : ab ∈ s), f ab =
  ∏ᶠ a b (h : b ∈ (s.filter (λ ab, prod.fst ab = a)).image prod.snd), f (a, b) :=
```
For example, given this we can easily prove:
```lean
@[to_additive] lemma finprod_mem_finset_of_product'₃ {γ : Type*}
  (s : finset (α × β × γ)) (f : α × β × γ → M) :
  ∏ᶠ abc (h : abc ∈ s), f abc = ∏ᶠ a b c (h : (a, b, c) ∈ s), f (a, b, c) :=
begin
  classical,
  rw finprod_mem_finset_of_product',
  simp_rw finprod_mem_finset_of_product',
  simp,
end
```

ocfnash replied:
I ended up adding various related bits and pieces inspired by your remarks. Not quite sure how much to include so any / all feedback most welcome!

ocfnash committed:
Add `finprod_curry`.

As it happens, I found it more convenient to derive this as a corollary
of `finprod_mem_finset_of_product` rather than vice-versa.

ocfnash committed:
Add `finprod_mem_finset_of_product'`

ocfnash committed:
Add `finprod_mem_finset_of_product₂`

ocfnash committed:
Add `finprod_curry₂`

eric-wieser commented on src/algebra/big_operators/finprod.lean:
```diff
@@ -205,6 +205,14 @@ open_locale big_operators
   ∏ᶠ (h : a ∈ s), f a = mul_indicator s f a :=
 by convert finprod_eq_if
 
+@[to_additive] lemma finprod_mem_mul_support (f : α → M) (a : α):
```
```suggestion
@[simp, to_additive] lemma finprod_mem_mul_support (f : α → M) (a : α) :
```

ocfnash replied:
Annoyingly I can't do this because `mem_mul_support` is tagged as `simp` :-/

eric-wieser replied:
Perhaps it would be better thhen to change the statement to `∏ᶠ (h : f a ≠ 1), f a = f a` then? Or do we already have that?

ocfnash replied:
We do not have that already but I think what we should really do is drop the `simp` on `mem_mul_support`.

I note also that even though I certainly can adjust `finprod_mem_mul_support` as you suggest (at the cost of sprinkling `mem_mul_support` in a few places) I cannot change the statement of `finprod_along_mul_support` to `∏ᶠ (f a ≠ 1), f a = ∏ᶠ a, f a`.

eric-wieser replied:
Sure, but you could change it to `∏ᶠ a (H : f a ≠ 1), f a = ∏ᶠ a, f a`. Remember that `∏ᶠ (a ∈ mul_support f), f a` is just shorthand for `∏ᶠ a (H : a ∈ mul_support f), f a`, they are indistinguishable to even lean meta code.

ocfnash replied:
Sorry my remark above was a bit of a thinko and isn't even relevent anyway. I think you're probably right about just restating this lemma as `∏ᶠ (h : f a ≠ 1), f a = f a` but I've added https://github.com/leanprover-community/mathlib/pull/7439/commits/db43cf0ad556118f50b87a83afce3624b2591771 as an experiment to see if it builds. I'll sort this out tomorrow.

ocfnash replied:
OK I think I've pushed up something sane now.

ocfnash committed:
Experiment with removing `simp` from `function.mem_mul_support`

We have to accept one of the following compromises:
  1. Not having `simp` for `finprod_mem_mul_support`
  2. Dropping `simp` from `function.mem_mul_support`
  3. Restating `finprod_mem_mul_support` using `h : f a ≠ 1` instead of `a ∈ mul_support f`

Probably 3 is the right approach but I'd at least like to see if the changes
here are sufficient to get the build passing so I can see what 2 would look like.

ocfnash committed:
Revert "Experiment with removing `simp` from `function.mem_mul_support`"

This reverts commit db43cf0ad556118f50b87a83afce3624b2591771.

ocfnash committed:
Add `simp` to `finprod_mem_mul_support`

Also dropping `finprod_along_mul_support` since this is now provable `by simp`.

ocfnash committed:
tweak

ocfnash committed:
merge master

ocfnash committed:
Tidy up whitespace

fpvandoorn commented on src/algebra/big_operators/finprod.lean:
```diff
@@ -700,6 +704,67 @@ lemma finsum_mul {R : Type*} [semiring R] (f : α → R) (r : R)
+
+@[to_additive] lemma mul_support_along_fiber_subset (f : α × β → M) (a : α) :
+  mul_support (λ b, f (a, b)) ⊆ (mul_support f).image prod.snd :=
+by tidy
```
This should be in `data.support`.

ocfnash replied:
Agreed, moved!

fpvandoorn reviewed (commented):
Some comments:
* all `..._of_product` names should be `..._product` or `..._prod` (that is the canonical name, but I guess that is confusing in this file).
* The subscripts `2` should probably be subscripts `3`, since they are about ternary functions.

ocfnash committed:
Naming suggestions from CR

ocfnash committed:
Suggestion from CR: move lemmas to `data/support.lean`
